Dockerfile

```diff
@@ -1,43 +1,90 @@
-# (unless we decide to use Bitnami instead)
-ARG base_image=ruby:3.0.4
+ARG ruby_version=3.0.4
+ARG base_image=ruby:$ruby_version-slim-bullseye
+ARG gem_home=/usr/local/bundle
 
 FROM $base_image AS builder
-# This image is only intended to be able to run this app in a production RAILS_ENV
-ENV RAILS_ENV=production
+SHELL ["/bin/bash", "-o", "pipefail", "-c"]
 
-# TODO: have a separate build image which already contains the build-only deps.
+ARG gem_home
+ENV RAILS_ENV=production \
+    RAILS_LOG_TO_STDOUT=1 \
+    NODE_ENV=production \
+    GEM_HOME=$gem_home \
+    BUNDLE_PATH=$gem_home \
+    BUNDLE_BIN=$gem_home/bin \
+    PATH=$gem_home/bin:$PATH \
+    BUNDLE_WITHOUT="development test cucumber" \
+    BOOTSNAP_CACHE_DIR=/var/cache/bootsnap
 
-# Add yarn to apt sources
-RUN curl -sS https://dl.yarnpkg.com/debian/pubkey.gpg | apt-key add -
-RUN echo "deb https://dl.yarnpkg.com/debian/ stable main" | tee /etc/apt/sources.list.d/yarn.list
+# TODO: set these in the builder image.
+ENV BUNDLE_IGNORE_MESSAGES=1 \
+    BUNDLE_SILENCE_ROOT_WARNING=1 \
+    BUNDLE_JOBS=12 \
+    MAKEFLAGS=-j12
 
+ENV ASSETS_PREFIX=/assets/whitehall \
+    GOVUK_APP_DOMAIN=unused \
+    GOVUK_WEBSITE_ROOT=unused
+
+# TODO: have an up-to-date builder image and stop running apt-get upgrade.
+# TODO: have a separate builder image which already contains the build-only deps.
 RUN apt-get update -qy && \
     apt-get upgrade -y && \
-    apt-get install -y build-essential nodejs yarn && \
-    apt-get clean
-RUN mkdir /app
+    apt-get install -y --no-install-recommends \
+        g++ make libc-dev curl yarnpkg libmariadb-dev-compat && \
+    ln -s /usr/bin/yarnpkg /usr/bin/yarn
+
+RUN mkdir -p /app && ln -fs /tmp /app/tmp && ln -fs /tmp /home/app
```

Comment:
why not make this part of the above RUN

Reply:
Could do, but it doesn't really belong there logically. The stuff above is about installing build-time dependencies via package managers, whereas this stanza is about building/installing the app.

Reply:
thats fair point